src/Hangfire.Tags.PostgreSql/TagsStorageTransactionRetryPolicy.cs

```diff
@@ -0,0 +1,43 @@
+using System;
+using System.Collections.Generic;
+using Hangfire.Logging;
+using Hangfire.Tags.Storage;
+using Npgsql;
+
+namespace Hangfire.Tags.PostgreSql
+{
+    /// <summary>
+    /// Retry policy to address https://github.com/face-it/Hangfire.Tags/issues/35
+    /// </summary>
+    public class TagsStorageTransactionRetryPolicy : ITagsStorageTransactionRetryPolicy
+    {
+        private readonly int _retryCount;
+        private readonly ILog _logger = LogProvider.For<TagsStorageTransactionRetryPolicy>(); 
+
+        public TagsStorageTransactionRetryPolicy(int retryCount)
+        {
+            _retryCount = retryCount;
+        }
+        
+        private static readonly HashSet<string> ConcurrencyErrorCodes = new HashSet<string>
+        {
+            "40001", //serialization_failure
+            "40P01"  //deadlock_detected 
+        };
+
+        public void RetryOnTransactionError(Action action)
+        {
+            for (int i = 1; i < _retryCount; i++)
+            {
+                try
+                {
+                    action();
+                }
+                catch (PostgresException e) when (ConcurrencyErrorCodes.Contains(e.SqlState))
+                {
+                    _logger.TraceException($"Operation failed with concurrency error. Retrying {i}/{_retryCount}", e);
```

Comment:
Should we leave logging here? I think, most of the time, `Trace` level is disabled in the user code. So it shouldn't put any stress on the user's logging environment.

But under high-load scenarios, there are tons of errors here. It may pollute logs if `Trace` level is enabled.

Reply:
Trace level might be disabled, but I do think the string format is still executed every time. I've added a check around that to prevent that from happening.